x/subscription/keeper/keeper.go

```diff
@@ -0,0 +1,405 @@
+package keeper
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"strconv"
+
+	"github.com/tendermint/tendermint/libs/log"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	paramtypes "github.com/cosmos/cosmos-sdk/x/params/types"
+	"github.com/crypto-org-chain/chain-main/v1/x/subscription/types"
+)
+
+type (
+	// Keeper of the chainmain store
+	Keeper struct {
+		cdc        codec.BinaryMarshaler
+		paramSpace paramtypes.Subspace
+		storeKey   sdk.StoreKey
+		sendKeeper types.SendKeeper
+	}
+)
+
+// NewKeeper creates a chainmain keeper
+func NewKeeper(
+	cdc codec.BinaryMarshaler, storeKey sdk.StoreKey, paramSpace paramtypes.Subspace,
+	sendKeeper types.SendKeeper,
+) Keeper {
+	// set KeyTable if it has not already been set
+	if !paramSpace.HasKeyTable() {
+		paramSpace = paramSpace.WithKeyTable(types.ParamKeyTable())
+	}
+	return Keeper{
+		cdc:        cdc,
+		paramSpace: paramSpace,
+		storeKey:   storeKey,
+		sendKeeper: sendKeeper,
+	}
+}
+
+// Logger returns a module-specific logger.
+func (k Keeper) Logger(ctx sdk.Context) log.Logger {
+	return ctx.Logger().With("module", fmt.Sprintf("x/%s", types.ModuleName))
+}
+
+func (k Keeper) SetParams(ctx sdk.Context, params *types.Params) {
+	k.paramSpace.SetParamSet(ctx, params)
+}
+
+func (k Keeper) GetParams(ctx sdk.Context, params *types.Params) {
+	k.paramSpace.GetParamSet(ctx, params)
+}
+
+// GetPlanID gets the highest plan ID
+func (k Keeper) GetPlanID(ctx sdk.Context) (planID uint64, err error) {
+	store := ctx.KVStore(k.storeKey)
+	bz := store.Get(types.PlanIDKey)
+	if bz == nil {
+		return 0, sdkerrors.Wrap(types.ErrInvalidGenesis, "initial plan ID hasn't been set")
+	}
+
+	planID = types.GetPlanIDFromBytes(bz)
+	return planID, nil
+}
+
+// SetPlanID sets the new plan ID to the store
+func (k Keeper) SetPlanID(ctx sdk.Context, planID uint64) {
+	store := ctx.KVStore(k.storeKey)
+	store.Set(types.PlanIDKey, types.GetPlanIDBytes(planID))
+}
+
+func (k Keeper) GetPlan(ctx sdk.Context, planID uint64) (types.Plan, bool) {
+	store := ctx.KVStore(k.storeKey)
+
+	bz := store.Get(types.PlanKey(planID))
+	if bz == nil {
+		return types.Plan{}, false
+	}
+
+	var plan types.Plan
+	k.MustUnmarshalPlan(bz, &plan)
+	return plan, true
+}
+
+func (k Keeper) SetPlan(ctx sdk.Context, plan types.Plan) {
+	store := ctx.KVStore(k.storeKey)
+	bz := k.MustMarshalPlan(plan)
+	store.Set(types.PlanKey(plan.PlanId), bz)
+}
+
+func (k Keeper) CreatePlan(ctx sdk.Context, msg types.MsgCreatePlan) (types.Plan, error) {
+	planID, err := k.GetPlanID(ctx)
+	if err != nil {
+		return types.Plan{}, err
+	}
+	plan := types.Plan{
+		Owner:        msg.Owner,
+		Title:        msg.Title,
+		Description:  msg.Description,
+		Price:        msg.Price,
+		PlanId:       planID,
+		DurationSecs: msg.DurationSecs,
+		CronSpec:     msg.CronSpec,
+	}
+
+	k.SetPlan(ctx, plan)
+	k.SetPlanID(ctx, planID+1)
```

Comment:
Why does the keeper always update planID to the latest planID + 1? What if we just store the ID of the latest plan in `SetPlan` and increase the ID counts only when creating the new plan?

Another question for the KVStore assign cost? Is it heavy to access KVStore by key? If the answer is yes, maybe we can try to reduce the KVStore assign across functions?

Reply:
It's used to generate unique plan id, the value stored in `PlanIdKey` is like the next plan id.

Reply:
> Another question for the KVStore assign cost? Is it heavy to access KVStore by key? If the answer is yes, maybe we can try to reduce the KVStore assign across functions?

Do you mean the cost of `ctx.KVStore(k.storeKey)`, and to share the returned store object?

Reply:
Yes. After checking your link. It looks good! Thanks.